### PR TITLE
Starting cpp version of PR026 test 

### PR DESCRIPTION
**Pull request by glass-ships**
Just started by duplicating `test_PR019_use_json_library.cpp`

### CODE REVIEW

**Comment by jpivarski** [comment]
Good start!

File: tests/test_PR030_recordarray_in_numba.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include "awkward/Slice.h"
+#include "awkward/fillable/FillableArray.h"
+#include "awkward/fillable/FillableOptions.h"
+
+namespace ak = awkward;
+
+int main(int, char**) 
+{
+  std::vector<std::vector<std::vector<double>>> vector =
+  {{{true, 1, 1.1}, {false, 2, 2.2}, {true, 3, 3.3}}, {}, {}, {}};
```
Comment (jpivarski) [inline tests/test_PR030_recordarray_in_numba.cpp]: This won't work: don't follow the original example this closely. It won't be easy to make C++ objects corresponding to the data (you'd have to define classes and all that), but you don't have to. Just call the appropriate `builder.XYZ` methods.

**Comment by jpivarski** [comment]
I'd like to approve your PR on a shorter timescale than I'll be able to update the README, and the QUICKSTART.md content you've written belongs in the REAME. When the time comes to merge this PR, it would be best if you remove QUICKSTART.md.

Also, note that your next step will be CMakeLists.txt, where you'll add your test. Probably the easiest way to get started is to make it a trivial test (`main` always returns `0`) to get the machinery in place, then add substantial tests, and make sure that doing it wrong causes the test to fail.

**Comment by jpivarski** [comment]
@glass-ships Tests are all green; you're good to go!

But try doing this:

```c++
myarray.beginrecord();
myarray.field("x");
myarray.boolean(true);
myarray.field("y");
myarray.integer(1);
myarray.field("z");
myarray.real(1.1);
myarray.endrecord();
```

instead to avoid creating unions. You probably want records that have [a field for each contained type](https://en.wikipedia.org/wiki/Product_type), not a union for which every [item is chosen from a set of types](https://en.wikipedia.org/wiki/Tagged_union).

**Comment by glass-ships** [comment]
Thanks for the tips, confirmed test is now working with your fix! I'll look at your suggested code, and keep in mind to avoid union types for the time being

**Comment by glass-ships** [comment]
I'm a little curious now that I'm switching the code around - what is the main difference between a "record" and a "list" in awkward's context? I think I understand the "field" object, it's a sort of organizational scheme paired with a value of a certain type, correct?

**Comment by jpivarski** [comment]
Think of this equivalence:

- Python list of all the same type of object ↔ Awkward list
- Python list of arbitrary types ↔ Awkward list of union (not implemented as an array yet)
- Python dict ↔ Awkward record
- one key in that dict ↔ Awkward field
- Python tuple of a fixed length ↔ Awkward tuple
- one slot in that tuple ↔ Awkward index
- quantity that could be `None` ↔ Awkward option (not implemented as an array yet)

Awkward arrays are a static type system (as are NumPy arrays, though NumPy is mostly limited to numeric types). In a static type system, a list has to contain objects of all the same type ("homogeneous") unless there's some structure that spells out what types are allowed: an Awkward union.

Python dicts are not statically typed; different dicts can have different keys and the values associated with those keys can be different types. Awkward records are more like C structs (or C++ classes). In an array with record type, all the records have to have the same keys and the values associated with one key have to all have the same type.

In Python, the distinction between lists and tuples is slight: tuples are immutable. In a static type system, there's more difference: lists in an array can have different lengths, but the items in those lists must have the same type; tuples in an array must have the same length, but the items in each slot can have different types. However, the items in a single slot `i` must have the same type for all tuples in the array.

In this kind of system, the difference between a tuple and a record is slight: both have a fixed number of fields that can have different types in the tuple/record (though each field must have the same type across the array), but the slots in a tuple are identified by integers while the slots in a record are identified by constant strings.

**Comment by glass-ships** [comment]
I thank you for the thorough explanation, that makes a lot of sense!

I've run into another problem, however, as it appears that:  
 `'class awkward::FillableArray' has no member named 'field'`.

 Which actually seems to check out, I couldn't find a member description in the FillableArray header or cpp files. It seems to work in the python version though, which makes me wonder, are the python and c++ functions defined separately? (it was my initial assumption the python was interfacing with awkward via c++ wrappers)

**Comment by jpivarski** [comment]
Oh—use `field_check`. That is what `field` is in Python.

The alternative, `field_fast`, doesn't check the string value, only its pointer, which makes it applicable only to strings that are known to have unchanging pointers (like literal strings, which is a common case). I also use `field_fast` in Numba, where I can create fixed pointers if the strings are known at compile-time.

If you're not sure, use `field_check`.

**Comment by jpivarski** [comment]
You can remove your QUICKSTART now; the information is on the new front-page README.

I'll help you with those conflicts soon.

**Comment by jpivarski** [comment]
I see you're online right now. Do you want me to merge with master, resolving all the conflicts (until the next PR...)?

**Comment by jpivarski** [comment]
@glass-ships The previous commit merges with master (takes all changes from master except your test). If it's not what you want, it can be reverted with git.

The next one fixes the test file so that it compiles (part of my sanity checks).

**Comment by glass-ships** [comment]
Sure! I think we're getting pretty close to wrapping this test up, too; once the array gets filled appropriately, I think we'll be happy with a simple print-out to json and confirm it matches what we expect it to.

**Comment by jpivarski** [comment]
I'm done merging now: `git pull`, do the "make clean" (that `rm` command, now on the front README), recompile, and you're ready to go!

**Comment by jpivarski** [comment]
Okay, now I'm done. I wanted to make sure that the only differences between this branch and master is in your test file. Now it is.

**Comment by jpivarski** [comment]
Any progress here?

You might want to fast-forward to the latest version of master, since the `FillableArray` → Awkward arrays conversion no longer has any unimplemented cases. Any `None` (`null`) values are represented by an `IndexedOptionArray` and any differences in types at the same level are represented by a `UnionArray`. If you had to do any work-arounds for missing types in the past, you no longer need to do them. Most of the unfinished bits are on the Python side, so you can write C++ unimpeded.

**Comment by glass-ships** [comment]
Thanks for pinging me on this, Jim! I kind of forgot to remove WIP from the title, but the autochecks seem happy, and test_PR026 seems to work with the most recent updates. What's your procedure for finalizing merge requests?

**Comment by jpivarski** [comment]
First, merge your branch to the latest master. That should trigger a test, but if it doesn't, I'll do that manually. Since you've given me the go-ahead, when the tests pass, I'll squash-and-merge it.

Keep in mind that a squash-and-merge is a lossy procedure: your local git repo will have the only record of the steps you took to complete the PR; GitHub will only have the before and after states. The PR will appear in the commit history as a single commit.

**Comment by jpivarski** [comment]
The branch was already up-to-date with master (aha! c205c991a22bab528b1e5e435029b1052a3f5c7e), so I'm just testing it again. If it passes, I'll merge it.

**Pull request merged**